content/languages/c/authoring/memory-management-techniques.md

```diff
@@ -6,19 +6,341 @@ sidebar: "language:c"
+    //allocate it once, and reuse it through the tests.
+    int* array = malloc(sizeof(int) * MAX_TEST);
+
+    //ten random tests
+    for(int i=0; i<10; ++i) {
+
+        //randomize the input
+        int n = rand() % MAX_TEST + 1;
+
+        //use preallocated array
+        calculate_numbers(n, array);
+
+        //...perform assertions, verify correctness of returned numbers...
+    }
+
+    //release the memory after all tests
+    free(array);
+}
+```
+
+
+#### When the size is not known, but is easy to calculate: ask and allocate
+
+Another possible approach is to ask user to implement _two_ functions: one which would return the size needed to fit the result, and another one to perform actual operation. For example:
+
+Kata task:
+
+> Given a positive integer `n`, return all terms of `n` top rows of Pascal's triangle, row by row, flattened into a single array.
+
+Solution:
+
+```c
+//Function which counts amount of terms to be returned
+int count_elements_of_pascal_triangle(int rows) {
+    return rows * (rows + 1) / 2;
+}
+
+//Function performing actual calculations
+void get_elements_of_pascal_triangle(int rows, int elements[]) {
+    //...calculate elements of Pascal's triangle and store them in elements array
+}
+```
+
+Tests:
+
+```c
+Test(fixed_tests, should_work_for_3) {
+
+    int rows = 3;
+    
+    //array allocated on stack,
+    //top three rows have 6 terms
+    int terms[6]; 
+
+    //pass the array to the function, and expect
+    //it to be filled with the result
+    get_elements_of_pascal_triangle(3, terms);
+
+    //...perform assertions, verify correctness of returned numbers...
+
+    //no need to deallocate the array
+}
+
+Test(random_tests, large_inputs) {
+
+    const int MAX_TEST = 1000;
+
+    //ten random tests
+    for(int i=0; i<10; ++i) {
+
+        //randomize the input
+        int rows = rand() % MAX_TEST + 1;
+
+        //query the solution for required size of the answer
+        int terms_count = count_elements_of_pascal_triangle(rows);
+
+        //You can perform assertions on the returned size here, or
+        //you can create a separate suite just to test the
+        //count_elements_of_pascal_triangle function
+
+        //dynamically allocate an array large enough to fit the answer
+        int* array = malloc(sizeof(int) * terms_count);
+
+        //use the allocated array when calling the user's solution
+        get_elements_of_pascal_triangle(rows, array);
+
+        //...perform assertions, verify correctness of returned numbers...
+
+        //release the memory after the test
+        free(array);
+    }
+}
+```
+
+This approach is used when the size of the answer cannot be easily inferred by the test suite, but can be efficiently calculated by the user, potentially without the overhead of calculating the actual solution.
+
+
+#### When the size is not known, and difficult to calculate: assume (or guess) the initial size and reallocate if too small
```

Comment:
This seem like a valid approach, but the fact that the size isn't easily predictable probably means that the task is relatively high-level and the solution is heavy enough that reallocating memory there isn't that bad, while recalculating the result after the buffer was too small is...

Reply:
This approach is definitely used, but in various variants. Works well for functions which just fetch data, but I've seen it also in APIs for digital signature and encryption. Callee can keep track of partial progress on their own and reuse it, or they can allocate memory locally, drive the process to its end, and upon return signal the required amount of memory by out param and copy potentially truncated amount of data to the output buffer, etc. Possibilities are many.

It's used by quite a number of Win32 API functions, for example for [reading from the Windows registry](https://docs.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-reggetvaluea) (in particular, see the note on `HKEY_PERFORMANCE_DATA`), but as it's said, characteristic of that function is _not_ to perform heavy computations.

Nonetheless, my intention is not to create C programmers guide, I am just looking for a way to transfer some ideas to C authors and translators, to get rid of ideas like requirement of mallocing a 3-element array to return hours, minutes, and seconds. Or returning `Position**`, where each entry is... a pointer to an individually malloced, single `Position` :( And there are even worse. I just thought about collecting some ideas and throw them at the authors, so they could pick something suitable for their kata. It's meant to be CW- and kata-oriented, and not an industry grade HOWTO ;) That's why I do not want to make things too complex and too detailed. Maybe I already have, maybe this paragraph, or this whole article, is simply not necessary?

Reply:
> if the whole point is to delegate allocations to the caller and it does not even know how the buffer was allocated

I feel like it's not too beneficial in this case. The reasons to do this may be:
* avoid reallocations if a function is called many times... it's not like it's called _millions_ of times on CW;
* allow using different allocation methods, especially in cross-platform code... is it worth it here?
So it's not like it's bad, but I'm not sure whether it should be in the tutorial. Those who are ready to read looong and complicated (relative to their level) tutorials may as well google themselves, while others would probably already be confused by this point. Anyway, I'm not the target audience, so I don't know what they need.

Reply:
Ok, I think I get what you mean, thanks. I will think what to do with this paragraph, but I think in the end I will just remove it.

Do you think other parts are more or less OK? Should be simplified? Or maybe the whole article is pointless as no one would benefit from it?

Reply:
I haven't read it all thoroughly, I'm not sure.

Reply:
This point has been removed.